User#followed_articles->Article.cached_tagged_with

jgaskins opened the pull request:
<!--
     For Work In Progress Pull Requests, please use the Draft PR feature,
     see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.

     For a timely review/response, please avoid force-pushing additional
     commits if your PR already received reviews or comments.

     Before submitting a Pull Request, please ensure you've done the following:
     - 📖 Read the Forem Contributing Guide: https://github.com/forem/forem/blob/master/CONTRIBUTING.md#create-a-pull-request.
     - 📖 Read the Forem Code of Conduct: https://github.com/forem/forem/blob/master/CODE_OF_CONDUCT.md.
     - 👷‍♀️ Create small PRs. In most cases this will be possible.
     - ✅ Provide tests for your changes.
     - 📝 Use descriptive commit messages.
     - 📗 Update any related documentation and include any relevant screenshots.

     NOTE: Pull Requests from forked repositories will need to be reviewed by
     a Forem Team member before any CI builds will run. Once your PR is approved
     with a `/ci` reply to the PR, it will be allowed to run subsequent builds without
     manual approval.
-->

## What type of PR is this? (check all applicable)

- [ ] Refactor
- [ ] Feature
- [ ] Bug Fix
- [x] Optimization
- [ ] Documentation Update

## Description

This should improve performance on StoriesController#index.

On DEV, the mismatch between `tagged_with` and `cached_tagged_with` is at most 0.007% — the `javascript` tag, with 4 mismatches, the last of which was published over 3 years ago.

Running it for all the tags I could think of off the top of my head, I came up with 0 mismatches on all articles published within the last year:

```ruby
articles = Article
  .published
  .where('published_at > ?', 1.year.ago)

%w[ruby javascript discuss meta elixir golang rust career performance].map do |tag|
  (articles.tagged_with(tag).count - articles.cached_tagged_with(tag).count).abs
end
# => [0, 0, 0, 0, 0, 0, 0, 0, 0]
```

## Added tests?

- [ ] Yes
- [x] No, and this is why: The existing tests should be sufficient to ensure the behavior has not changed
- [ ] I need help with writing tests

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [ ] I've updated the [Developer Docs](https://docs.forem.com) and/or
      [Admin Guide](https://forem.gitbook.io/forem-admin-guide/), or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
- [ ] I've updated the README or added inline documentation
- [ ] I will share this change in a [Changelog](https://forem.dev/t/changelog)
      or in a [forem.dev](http://forem.dev) post
- [x] I will share this change internally with the appropriate teams
- [ ] I'm not sure how best to communicate this change and need help
- [ ] This change does not need to be communicated, and this is why not: _please
      replace this line with details on why this change doesn't need to be
      shared_

jgaskins commented:
Heads up @rhymes, this goes along with what we've been discussing about querying on the denormalized column to improve query times. I pulled some of the data and this honestly looks pretty safe.

_Note:_ I'm not looking to deploy it this week until we can get a baseline before optimizing the `EmailDigestArticleCollector`, which won't be until Friday (it [won't run until Wednesday](https://github.com/forem/forem/blob/624c671fa48b6916345ff23abe0a7c25a0ec11a1/lib/tasks/fetch.rake#L5) and I'd like to get a few days of traces on it), but I definitely wanted to get this information in front of you in case it helps for later — we may be able to migrate _all_ of the `Article.tagged_with` to `Article.cached_tagged_with` and speed up these queries for free due to the index you've already added [here](https://github.com/forem/forem/pull/13052/files#diff-4d61a839368e72c9b246744c46530276c5b9963c60bd2f3e139af3e535c54b40R7).

I still need to run some benchmarks to make sure these are actually faster for these queries, but it should be able to use a whole lot of indexes that already exist on the `articles` table with this change.

rhymes commented:
@jgaskins good plan, one tiny defailt: as far as I know `.tagged_with` works with multiple tags (hence the `any: true` in that call):

```ruby
[2] pry(main)> Article.tagged_with(%w[ruby python]).to_sql
=> "SELECT \"articles\".* FROM \"articles\" INNER JOIN \"taggings\" \"article_taggings_18e40e1\" ON \"article_taggings_18e40e1\".\"taggable_id\" = \"articles\".\"id\" AND \"article_taggings_18e40e1\".\"taggable_type\" = 'Article' AND \"article_taggings_18e40e1\".\"tag_id\" IN (SELECT \"tags\".\"id\" FROM \"tags\" WHERE \"tags\".\"name\" LIKE 'ruby' ESCAPE '!') INNER JOIN \"taggings\" \"article_taggings_4235227\" ON \"article_taggings_4235227\".\"taggable_id\" = \"articles\".\"id\" AND \"article_taggings_4235227\".\"taggable_type\" = 'Article' AND \"article_taggings_4235227\".\"tag_id\" IN (SELECT \"tags\".\"id\" FROM \"tags\" WHERE \"tags\".\"name\" LIKE 'python' ESCAPE '!')"
[3] pry(main)> Article.tagged_with(%w[ruby python], any: true).to_sql
=> "SELECT \"articles\".* FROM \"articles\" WHERE EXISTS (SELECT * FROM \"taggings\" WHERE \"taggings\".\"taggable_id\" = \"articles\".\"id\" AND \"taggings\".\"taggable_type\" = 'Article' AND \"taggings\".\"tag_id\" IN (SELECT \"tags\".\"id\" FROM \"tags\" WHERE (\"tags\".\"name\" LIKE 'ruby' ESCAPE '!' OR \"tags\".\"name\" LIKE 'python' ESCAPE '!')))"
[4] pry(main)> Article.cached_tagged_with(%w[ruby python]).to_sql
=> "SELECT \"articles\".* FROM \"articles\" WHERE (cached_tag_list ~* '^[\"ruby\", \"python\"],| [\"ruby\", \"python\"],|, [\"ruby\", \"python\"]$|^[\"ruby\", \"python\"]$')"
[5] pry(main)> Article.cached_tagged_with(%w[ruby python]).first
  Article Load (71.2ms)  SELECT "articles".* FROM "articles" WHERE (cached_tag_list ~* '^["ruby", "python"],| ["ruby", "python"],|, ["ruby", "python"]$|^["ruby", "python"]$') ORDER BY "articles"."id" ASC LIMIT $1  [["LIMIT", 1]]
=> nil
```

you can see it's not working correctly with multiple tags:

```ruby
[6] pry(main)> Article.cached_tagged_with(%w[ruby python]).count
   (25.7ms)  SELECT COUNT(*) FROM "articles" WHERE (cached_tag_list ~* '^["ruby", "python"],| ["ruby", "python"],|, ["ruby", "python"]$|^["ruby", "python"]$')
=> 0
[7] pry(main)> Article.tagged_with(%w[ruby python], any: true).count
   (128.4ms)  SELECT COUNT("articles".*) FROM "articles" WHERE EXISTS (SELECT * FROM "taggings" WHERE "taggings"."taggable_id" = "articles"."id" AND "taggings"."taggable_type" = 'Article' AND "taggings"."tag_id" IN (SELECT "tags"."id" FROM "tags" WHERE ("tags"."name" LIKE 'ruby' ESCAPE '!' OR "tags"."name" LIKE 'python' ESCAPE '!')))
=> 2277
[8] pry(main)> Article.tagged_with(%w[ruby python]).count
   (30.6ms)  SELECT COUNT(*) FROM "articles" INNER JOIN "taggings" "article_taggings_18e40e1" ON "article_taggings_18e40e1"."taggable_id" = "articles"."id" AND "article_taggings_18e40e1"."taggable_type" = 'Article' AND "article_taggings_18e40e1"."tag_id" IN (SELECT "tags"."id" FROM "tags" WHERE "tags"."name" LIKE 'ruby' ESCAPE '!') INNER JOIN "taggings" "article_taggings_4235227" ON "article_taggings_4235227"."taggable_id" = "articles"."id" AND "article_taggings_4235227"."taggable_type" = 'Article' AND "article_taggings_4235227"."tag_id" IN (SELECT "tags"."id" FROM "tags" WHERE "tags"."name" LIKE 'python' ESCAPE '!')
=> 1
```

I think we should expand `.cached_tagged_with` to be smarter :)

jgaskins commented:
@rhymes 

> good plan, one tiny defailt: as far as I know `.tagged_with` works with multiple tags (hence the `any: true` in that call)

Ooh, good point. I didn't think about multiple tags. I'll add some tests for that.
 
> I think we should expand `.cached_tagged_with` to be smarter :)

I have some plans for that scope, too! :-) I actually simplified it a bit locally but I didn't commit it because I wanted to keep this PR focused, but if this should accept multiple tags to match `tagged_with` maybe I should just go ahead with it. Do you have any thoughts on that?

jgaskins commented:
@rhymes In 7f2fd4a812796f52e53ab347b188f4c30a01b486 I updated `cached_tagged_with` and added `cached_tagged_with_any`, including specs for both — I couldn't find any tests for `cached_tagged_with`, so I added a `describe` block for it. In the specs for both, I compared the output to the output of `tagged_with` so we could check its results against the slower but canonical version.

I still need to update `User#followed_articles` to use that `_any` scope since it was using `any: true` before.

jgaskins commented:
Just as a quick example of the performance difference on `EmailDigestArticleCollector#articles_to_send`, here is what the current query looks like on a prod DB snapshot for Molly's feed (she follows quite a few tags and people so her feed takes a while to populate), taking almost 600ms. The first run took > 7 seconds but I'm assuming that's simply due to not having the data in the DB's in-memory cache yet):

```
  Article Load (591.9ms)  SELECT "title", "description", "path" FROM ( (SELECT "articles".* FROM "articles" WHERE EXISTS (SELECT * FROM "taggings" WHERE "taggings"."taggable_id" = "articles"."id" AND "taggings"."taggable_type" = 'Article' AND "taggings"."tag_id" IN (SELECT "tags"."id" FROM "tags" WHERE ("tags"."name" LIKE 'requestforpost' ESCAPE '!' OR "tags"."name" LIKE 'discuss' ESCAPE '!' OR "tags"."name" LIKE 'productivity' ESCAPE '!' OR "tags"."name" LIKE 'ruby' ESCAPE '!' OR "tags"."name" LIKE 'jokes' ESCAPE '!' OR "tags"."name" LIKE 'beginners' ESCAPE '!' OR "tags"."name" LIKE 'security' ESCAPE '!' OR "tags"."name" LIKE 'career' ESCAPE '!')))) UNION (SELECT "articles".* FROM "articles" WHERE "articles"."user_id" IN ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16, $17, $18, $19, $20, $21, $22, $23, $24, $25, $26, $27, $28, $29, $30, $31, $32, $33, $34, $35, $36, $37, $38, $39, $40, $41, $42, $43, $44, $45, $46, $47, $48, $49, $50, $51, $52, $53, $54, $55, $56, $57, $58, $59, $60, $61, $62, $63, $64, $65, $66, $67, $68, $69, $70, $71, $72, $73, $74, $75, $76, $77, $78, $79, $80, $81, $82, $83, $84, $85, $86, $87, $88, $89, $90, $91, $92, $93, $94, $95, $96, $97, $98, $99, $100, $101, $102, $103, $104, $105, $106, $107, $108, $109, $110, $111, $112, $113, $114, $115, $116, $117, $118, $119, $120, $121, $122, $123, $124, $125, $126, $127, $128, $129, $130, $131, $132, $133, $134, $135, $136, $137, $138, $139, $140, $141, $142, $143, $144, $145, $146, $147, $148, $149, $150)) ) "articles" WHERE "articles"."published" = $151 AND (published_at <= '2021-03-30 16:49:04.917569') AND (published_at > '2021-03-02 16:49:04.917730') AND "articles"."email_digest_eligible" = $152 AND "articles"."user_id" != $153 AND (score > 12) AND (experience_level_rating > 6.4 AND experience_level_rating < 13.6) ORDER BY "score" DESC LIMIT $154
```

The new query using `cached_tagged_with_any` instead of `tagged_with(_, any: true)`, 20x as fast:

```
  Article Load (29.0ms)  SELECT "title", "description", "path" FROM ( (SELECT "articles".* FROM "articles" WHERE ((((((((cached_tag_list ~ '[[:<:]]requestforpost[[:>:]]') OR (cached_tag_list ~ '[[:<:]]discuss[[:>:]]')) OR (cached_tag_list ~ '[[:<:]]productivity[[:>:]]')) OR (cached_tag_list ~ '[[:<:]]ruby[[:>:]]')) OR (cached_tag_list ~ '[[:<:]]jokes[[:>:]]')) OR (cached_tag_list ~ '[[:<:]]beginners[[:>:]]')) OR (cached_tag_list ~ '[[:<:]]security[[:>:]]')) OR (cached_tag_list ~ '[[:<:]]career[[:>:]]'))) UNION (SELECT "articles".* FROM "articles" WHERE "articles"."user_id" IN ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16, $17, $18, $19, $20, $21, $22, $23, $24, $25, $26, $27, $28, $29, $30, $31, $32, $33, $34, $35, $36, $37, $38, $39, $40, $41, $42, $43, $44, $45, $46, $47, $48, $49, $50, $51, $52, $53, $54, $55, $56, $57, $58, $59, $60, $61, $62, $63, $64, $65, $66, $67, $68, $69, $70, $71, $72, $73, $74, $75, $76, $77, $78, $79, $80, $81, $82, $83, $84, $85, $86, $87, $88, $89, $90, $91, $92, $93, $94, $95, $96, $97, $98, $99, $100, $101, $102, $103, $104, $105, $106, $107, $108, $109, $110, $111, $112, $113, $114, $115, $116, $117, $118, $119, $120, $121, $122, $123, $124, $125, $126, $127, $128, $129, $130, $131, $132, $133, $134, $135, $136, $137, $138, $139, $140, $141, $142, $143, $144, $145, $146, $147, $148, $149, $150)) ) "articles" WHERE "articles"."published" = $151 AND (published_at <= '2021-03-30 16:52:58.594281') AND (published_at > '2021-03-02 16:52:58.594430') AND "articles"."email_digest_eligible" = $152 AND "articles"."user_id" != $153 AND (score > 12) AND (experience_level_rating > 6.4 AND experience_level_rating < 13.6) ORDER BY "score" DESC LIMIT $154
```

The result sets for both queries are the same.

rhymes commented on app/models/article.rb:
```diff
@@ -146,7 +146,33 @@ class Article < ApplicationRecord
+  scope :cached_tagged_with, lambda { |tag|
+    case tag
+    when String
+      where("cached_tag_list ~ ?", "[[:<:]]#{tag}[[:>:]]")
```
Could you somehow explain `"[[:<:]]#{tag}[[:>:]]"` within a comment? I think this is arcane enough for a comment :D. I found this in PostgreSQL documentation:

> There are two special cases of bracket expressions: the bracket expressions [[:<:]] and [[:>:]] are constraints, matching empty strings at the beginning and end of a word respectively

rhymes commented:
@jgaskins great work! 😭 - I have no suggestions ahhaha :D

jgaskins commented:
@rhymes Good call on adding the comment! I'll go ahead and do that.

In the meantime, [this](https://app.datadoghq.com/apm/resource/practical_developer/EmailDigestArticleCollector_articles_to_send/69c3eb25cf0abd0a?end=1617222347335&env=production&paused=true&query=env%3Aproduction%20service%3Apractical_developer%20operation_name%3AEmailDigestArticleCollector_articles_to_send%20resource_name%3AEmailDigestArticleCollector%23articles_to_send&spanID=3196947466819692374&start=1617189821461) is what the performance looked like on DEV for the today.

From an arbitrary user in that list:

<img width="275" alt="Time spent in SQL query: 947ms" src="https://user-images.githubusercontent.com/108205/113212660-10f7b500-9245-11eb-9a3e-5e74400cd7b3.png">

I opened a prod Rails console and monkeypatched the scopes and the new query for the same user ends up coming down from 947ms to 47ms, a reduction in latency of 95%:

```
D, [2021-03-31T21:18:39.386718 #3] DEBUG -- :   Article Load (47.3ms)  SELECT "title", "description", "path" FROM ( (SELECT "articles".* FROM "articles" WHERE (((((((((((((((((((((((((cached_tag_list ~ '[[:<:]]git[[:>:]]') OR (cached_tag_list ~ '[[:<:]]react[[:>:]]')) OR (cached_tag_list ~ '[[:<:]]c[[:>:]]')) OR (cached_tag_list ~ '[[:<:]]javascript[[:>:]]')) OR (cached_tag_list ~ '[[:<:]]django[[:>:]]')) OR (cached_tag_list ~ '[[:<:]]bash[[:>:]]')) OR (cached_tag_list ~ '[[:<:]]laravel[[:>:]]')) OR (cached_tag_list ~ '[[:<:]]typescript[[:>:]]')) OR (cached_tag_list ~ '[[:<:]]devops[[:>:]]')) OR (cached_tag_list ~ '[[:<:]]java[[:>:]]')) OR (cached_tag_list ~ '[[:<:]]hacktoberfest[[:>:]]')) OR (cached_tag_list ~ '[[:<:]]python[[:>:]]')) OR (cached_tag_list ~ '[[:<:]]docker[[:>:]]')) OR (cached_tag_list ~ '[[:<:]]machinelearning[[:>:]]')) OR (cached_tag_list ~ '[[:<:]]vue[[:>:]]')) OR (cached_tag_list ~ '[[:<:]]go[[:>:]]')) OR (cached_tag_list ~ '[[:<:]]linux[[:>:]]')) OR (cached_tag_list ~ '[[:<:]]ruby[[:>:]]')) OR (cached_tag_list ~ '[[:<:]]vim[[:>:]]')) OR (cached_tag_list ~ '[[:<:]]cloud[[:>:]]')) OR (cached_tag_list ~ '[[:<:]]rust[[:>:]]')) OR (cached_tag_list ~ '[[:<:]]angular[[:>:]]')) OR (cached_tag_list ~ '[[:<:]]beginners[[:>:]]')) OR (cached_tag_list ~ '[[:<:]]webdev[[:>:]]')) OR (cached_tag_list ~ '[[:<:]]opensource[[:>:]]'))) UNION (SELECT "articles".* FROM "articles" WHERE "articles"."user_id" IN ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12)) ) "articles" WHERE "articles"."published" = $13 AND (published_at <= '2021-03-31 21:18:39.337349') AND (published_at > '2021-03-27 21:18:39.330071') AND "articles"."email_digest_eligible" = $14 AND "articles"."user_id" != $15 AND (score > 12) AND (experience_level_rating > 1.4 AND experience_level_rating < 8.6) ORDER BY "score" DESC LIMIT $16
```

[The benchmark I ran yesterday](https://github.com/forem/forem/pull/13174#issuecomment-810442758) was against a prod DB snapshot on my own machine and I'm getting almost the exact same ratio against the actual prod DB from a prod console, which is a great sign.

maestromac reviewed (approved):
:heart_eyes:  recursion. LGTM!

rhymes reviewed (approved):
Forgot to approve :D 

Great numbers @jgaskins!

jgaskins commented:
With this merged, [this latency graph](https://app.datadoghq.com/apm/resource/practical_developer/EmailDigestArticleCollector_articles_to_send/69c3eb25cf0abd0a?end=1617308601205&env=production&fullscreen_end_ts=1617308632988&fullscreen_paused=false&fullscreen_section=overview&fullscreen_start_ts=1617135832988&paused=false&query=env%3Aproduction%20service%3Apractical_developer%20operation_name%3AEmailDigestArticleCollector_articles_to_send%20resource_name%3AEmailDigestArticleCollector%23articles_to_send&spanID=2532813318566482201&start=1617135801205) should plummet to almost nothing tomorrow morning. If that theory holds, we can remove [this check](https://github.com/forem/forem/blob/9533c77d7f038bff1697aa748ca874df736e3768/app/workers/emails/enqueue_digest_worker.rb#L8-L11) and get perf data on that GitHub worker and see the performance impact on database queries in general.

That last part will happen next week at the earliest. I'll be on call over the weekend but I'm not going to give myself a reason to work the weekend. 😄 

_Less important stuff below this line_

---

If it puts too much stress on the DEV DB, we can reschedule the email-digest Sidekiq job for a time when we're receiving less traffic. Our traffic dips every evening and this job is scheduled to run at a time when our traffic is [already over halfway between the local minimum and the local maximum](https://app.datadoghq.com/dashboard/7w7-zg8-rnz?from_ts=1617115527548&fullscreen_end_ts=1617311103522&fullscreen_paused=true&fullscreen_section=overview&fullscreen_start_ts=1617138303522&fullscreen_widget=1497562214493233&live=true&tile_size=m&to_ts=1617288327548), so it might be worth rescheduling that job to run later in the day if it puts too much pressure on the DB.

It may be worth making the start time of that job configurable so that other high-volume Forem instances with very different traffic patterns can run it at a different time. I'm not suggesting we do that now because we don't seem to have any Forem instances with that much volume yet — just saying it out loud in case it becomes a thing.
